File: solutions/Matrix-Search.cpp
int nLessThanOrEqualTo(int target, vector<vector<int>> &matrix)
{
    int res = 0;
    for (auto row : matrix)
        res += std::upper_bound(row.begin(), row.end(), target) - row.begin();
    return res;
}

int solve(vector<vector<int>> &matrix, int k)
{
    int lo = matrix[0][0];
    int hi = matrix.back().back();

    while (lo < hi)
    {
        int mid = (lo + hi) / 2;
        // find how many elements in matrix that's <= `mid`
        // if this number is > k, search left
        // otherwise search right
        if (nLessThanOrEqualTo(mid, matrix) > k)
            hi = mid;
        else
            lo = mid + 1;
    }
    return lo;
}

File: solutions/Matrix-Search.py
from heapq import heappush, heappop


class Solution:
    def solve(self, matrix, k):
        # use a heap to store k + 1 smallest elements at any time
        heap = []
        for row in matrix:
            for val in row:
                heappush(heap, -val)
                if len(heap) > k + 1:
                    heappop(heap)

        return -heappop(heap)
